build-tools/packages/build-tools/src/type-test-generator/typeTestUtils.ts

```diff
@@ -0,0 +1,289 @@
+ */
+export function getTypeDefinitionFilePath(previousBasePath: string): string {
+	const previousPackageJson: PackageJson = readJsonSync(
+		getPreviousPackageJsonPath(previousBasePath),
+	);
+	// Check the exports entries
+	if (previousPackageJson.exports) {
+		return getTypePathFromExport(previousPackageJson, previousBasePath);
+		// Check the types field from the previous package.json as a fallback
+	} else if (previousPackageJson.types) {
+		return path.join(previousBasePath, previousPackageJson.types);
+	} else {
+		throw new Error(
+			`Type definition file path could not be determined from '${getPreviousPackageJsonPath(
+				previousBasePath,
+			)}'. No 'exports' nor 'type' fields found.`,
+		);
+	}
+}
+
+/**
+ * Extracts type data from a TS source file and creates a map where each key is a type name and the value is its type data.
+ * @param file - The source code file containing type data
+ * @returns The mapping between item and its type
+ */
+export function typeDataFromFile(file: SourceFile): Map<string, TypeData> {
+	const typeData = new Map<string, TypeData>();
+	const exportedDeclarations = file.getExportedDeclarations();
+
+	for (const declarations of exportedDeclarations.values()) {
+		for (const dec of declarations) {
+			getNodeTypeData(dec).forEach((td) => {
+				const fullName = getFullTypeName(td);
+				if (typeData.has(fullName)) {
+					// This system does not properly handle overloads: instead it only keeps the last signature.
+					console.warn(`skipping overload for ${fullName}`);
+				}
+				typeData.set(fullName, td);
+			});
+		}
+	}
+	return typeData;
+}
+
+/**
+ * Initializes TypeScript projects for the current and previous package versions and loads specific source files.
+ * @param {string} typeDefinitionFilePath - The path to the type definition file for the previous version.
+ * @returns {{ currentFile: SourceFile, previousFile: SourceFile }} - The loaded source files for the current and previous versions.
+ */
+export function initializeProjectsAndLoadFiles(
+	typeDefinitionFilePath: string,
+	previousBasePath: string,
+	previousPackageName: string,
+) {
+	const currentFile = new Project({
+		skipFileDependencyResolution: true,
+		tsConfigFilePath: "tsconfig.json",
+	}).getSourceFileOrThrow("index.ts");
+
+	const previousTsConfigPath = path.join(previousBasePath, "tsconfig.json");
+	const project = new Project({
+		skipFileDependencyResolution: true,
+		tsConfigFilePath: existsSync(previousTsConfigPath) ? previousTsConfigPath : undefined,
+	});
+
+	let previousFile: SourceFile;
+	// Check for existence of alpha and add appropriate file
+	if (existsSync(typeDefinitionFilePath)) {
+		project.addSourceFilesAtPaths(typeDefinitionFilePath);
+		previousFile = project.getSourceFileOrThrow(`${previousPackageName}-alpha.d.ts`);
+		// Fall back to using .d.ts
+	} else {
+		project.addSourceFilesAtPaths(`${previousBasePath}/dist/**/*.d.ts`);
+		previousFile = project.getSourceFileOrThrow("index.d.ts");
+	}
```

Comment:
@alexvy86 Yes, the logic is different because it was updated to first attempt to use the /alpha path for type definitions and then fall back to the dist folder's,  `.d.ts` files if the /alpha path is not available.

The logic above checks if typeDefinitionFilePath exists, which implies the existence of `alpha` path for type definitions. If it doesn't exist,  he code falls back to using .d.ts files located at ${previousBasePath}/dist/**/*.d.ts. In this case, previousFile is set to the source file named "index.d.ts".

Reply:
My main question is if using `index.ts` (not `index.d.ts`) before was a mistake? Or why do we not try to do that at all anymore?

Reply:
@alexvy86, I think my change was unintentional, but correct because using the `index.d.ts` file implies that you are looking for explicit type declarations that are separate from the implementation in index.ts

We generally don't have any implementation details in index.ts so I'm not sure if its initial use was a mistake but I'd say it feels more appropriate to use index.d.ts as the fallback because as a declaration file, it contains only type declarations and not any executable code. What do you think?

CC: @tylerbutler

Reply:
@alexvy86, I think my change was unintentional, but correct because using the `index.d.ts` file implies that you are looking for explicit type declarations that are separate from the implementation in index.ts

We generally don't have any implementation details in index.ts so I'm not sure if its initial use was a mistake but I'd say it feels more appropriate to use index.d.ts as the fallback because as a declaration file, it contains only type declarations and not any executable code. What do you think?

CC: @tylerbutler